### ISSUE TITLE
Feature: Simulate Pods scheduling process in any existing clusters

### ISSUE THREAD

**Issue by lianghao208** [kubernetes-sigs/kube-scheduler-simulator#41]
Hi, I am new to this project (from scheduling-plugin sig). I learned about the kube-scheduler-simulator. In my option, this simulator only simulate pods scheduling process in a new simulation cluster, which means it can't simulate a real cluster for now(e.g production environment) . Is it possible to simulate pods in an already exist environment?
I am interesting in contributing to this project, maybe I need to learn more details about it.

**Comment by sanposhiho** [comment]
Hi @lianghao208 :)

Currently, when you start the simulator, there are no resources on the simulator, and we have to create resources(pods, nodes, ...) one by one to simulate cluster. 

#28 is one idea to enable user to export current resources **on the simulator** to some file and import resources with that file later. 
But your idea is importing resources from **a real cluster**, right? That makes sense to me.

> I am interesting in contributing to this project, maybe I need to learn more details about it.

Yes, you can work on this feature if you want. I will support you as much as I can.

**Comment by lianghao208** [comment]
> But your idea is importing resources from a real cluster, right? That makes sense to me :)

Yes, my idea is to import a real cluster's etcd data (etcd's endpoint, certs are needed) into the simulator, then the simulator will create the same resources as the imported cluster's.

> Yes, you can work on this feature if you want. I will support you as much as I can.

Thank you, I will try to work on this.

**Comment by sanposhiho** [comment]
> my idea is to import a real cluster's etcd data (etcd's endpoint, certs are needed) into the simulator, then the simulator will create the same resources as the imported cluster's.

I see.

Let me ask you one question. Can't most users get certs of etcd easily because getting etcd certs requires strong permissions in the cluster? Do you expect the administrator to use this export feature? Or is there an easy way for users to get the cert for the real cluster etcd? (I'm not very familiar with etcd certificate in Kubernetes.)

I think there are several ways to achieve this import feature, including your way. And the way I'm thinking is to get resources from api-server of real cluster. 

> I will try to work on this.

Thanks. Feel free to `/assign` :)

**Comment by flowscolors** [comment]
Hello, thank you very much for this project, I look forward to it very much. We also need to check the scheduling situation on a real kubernetes cluster. We operate and maintain about 60+ kuebrnetes clusters. As a cluster administrator, users often ask why two Pods are scheduled to one node, which causes container exceptions, such as local PV cannot be allocated. We can only increase the scheduler's log level  to 9 to view the log. We hope this project can help us. And In this way，I think we should get information from apiserver instead of etcd. After all, providing etcd's certificate means that there is one more entry for the cluster to be destroyed. And Apiserver can use webhook to integrate various authentication channels.

**Comment by sanposhiho** [comment]
By #64, we created a new service to export/import all resources. 
I think we can use that to achieve this.

0. users passes information to connect kube-apiserver in existing cluster.
1. When simulator is start, fetch all needed resources from existing cluster (via kube-apiserver)
2. import that resources via `Import` func ([ref](https://github.com/kubernetes-sigs/kube-scheduler-simulator/blob/master/export/export.go#L155))

should be enough, maybe?

**Comment by 196Ikuchil** [comment]
Hi👋
 I would like to work on this issue, ok?

**Comment by sanposhiho** [comment]
@196Ikuchil 
Yes. I think it should be Okay.

(@lianghao208 Sorry if you've already worked on this. Thanks for proposing this nice feature.)

**Comment by lianghao208** [comment]
@sanposhiho It's ok, I think @196Ikuchil will have a better solution. I can review @196Ikuchil 's PR after it has been submitted.  

> - users passes information to connect kube-apiserver in existing cluster.
> - When simulator is start, fetch all needed resources from existing cluster (via kube-apiserver)
> - import that resources via Import func ([ref](https://github.com/kubernetes-sigs/kube-scheduler-simulator/blob/master/export/export.go#L155))


If the whole process can be separated into multiple tasks, I would like to work on some of them too.

**Comment by sanposhiho** [comment]
@lianghao208 

>  I can review @196Ikuchil 's PR after it has been submitted.

Thanks. 
Since this feature enables the communication between real-world Kubernetes and the simulator and I'm not very experienced as a Kubernetes operator, having a reviewer other than myself would be very helpful. 

---

> If the whole process can be separated into multiple tasks, I would like to work on some of them too.

Yeah. In detail, I think the following changes are needed. 
Please let me know if there are any concerns, lack of explanation, or other good ideas.

- create an option to enable this import from existing cluster feature. 
  - All options for the simulator are located here: https://github.com/kubernetes-sigs/kube-scheduler-simulator/blob/master/config/config.go
- add `ignore error option` in `Export` and `Import` method in [ExportService](https://github.com/kubernetes-sigs/kube-scheduler-simulator/blob/1c9c6319b0140a0cc913d199f4fcd558c09478df/export/export.go). 
  When `ignore error option` is passed, `Export` method shouldn't return errors and log the error instead. 
  - The simulator may only be able to get some of the resources due to lack of permissions or other reasons. 
  - And some users may want to import just some of the resources from an existing cluster.
  - So, use this `ignore error option` when getting resources from existing cluster.
- create CopyResourcesService (tentative name) and all logic for this feature should be in that service.
  - CopyResourcesService will have two ExportService; one is for api-server in the simulator, the another is for existing cluster.
  - CopyResourcesService fetches all resources from existing cluster via Export method in ExportService for existing cluster. and Import them via Import method in ExportService for api-server in the simulator.
  - CopyResourcesService should be crated only when this feature is enabled by the option.
    - To create ExportService for the existing cluster, we need to create client from .kubeconfig file.
      - ExportService for the existing cluster requires creating other services for existing cluster as well.
- documentation for this feature.

@lianghao208 @196Ikuchil 
If we can reach an agreement on this, I think we can proceed with development in sections. (Then let's discuss and decide who will be in charge of changing which parts of this feature.)

**Comment by 196Ikuchil** [comment]
@sanposhiho 
I generally agree.

I have implemented the `ExportService`, so I should be in charge of this at least.
> add ignore error option in Export and Import method in [ExportService](https://github.com/kubernetes-sigs/kube-scheduler-simulator/blob/1c9c6319b0140a0cc913d199f4fcd558c09478df/export/export.go).

**Comment by 196Ikuchil** [comment]
@sanposhiho 
Hi, I'm working on the second issue.

Where should we pass the ignore option? By `ExportService` or each method calling?
```go
type Service struct {
	client               clientset.Interface
	podService           PodService
	nodeService          NodeService
	pvService            PersistentVolumeService
	pvcService           PersistentVolumeClaimService
	storageClassService  StorageClassService
	priorityclassService PriorityClassService
	schedulerService     SchedulerService
	ignoreErr            bool // this
}
...
func NewExportService(client clientset.Interface, pods PodService, nodes NodeService, pvs PersistentVolumeService, pvcs PersistentVolumeClaimService, storageClasss StorageClassService, priorityClasss PriorityClassService, schedulers SchedulerService, ignoreErr bool) *Service {
....
```
[source code](https://github.com/kubernetes-sigs/kube-scheduler-simulator/blob/17382c9fea30492335e19796841e14b695a4aaf9/export/export.go#L28)

OR
```go
func (s *Service) Export(ctx context.Context, ignoreErr bool) (*ResourcesForExport, error) {
...
```
[source code](https://github.com/kubernetes-sigs/kube-scheduler-simulator/blob/17382c9fea30492335e19796841e14b695a4aaf9/export/export.go#L143)

The first one looks better for me because this `ignore error option` is clearly determined by whether it is for local or external, I think.

**Comment by 196Ikuchil** [comment]
I think this point also is not determined.
**When should we pass the config file (like a `kubeconfig`)?** 

**Is it when we launch the scheduler on the terminal? (e.g. environment variable)**
-  there is a possible config file that also has already been prepared if someone does set `ignore error option` to true before launch.
- config file is imported directly. (Not through the network.) This is more secure, If the schedular work on an external server.

**Or, is it via a web form?**
- it is easy to manipulate for any user.
- It is easy to reupload(reimport) another config file.

**Comment by sanposhiho** [comment]
> Where should we pass the ignore option? By each ExportService or each method calling?

I prefer each method call. And it should be implemented as functional option pattern.
https://github.com/uber-go/guide/blob/master/style.md#functional-options

> The first one looks better for me because this ignore error option is clearly determined by whether it is for local or external, I think.

-1. We may use it in different places in the future. We can never fully know how it will be used in the future.
Since the number of methods is not very large, it is better to implement options for each function, so that they can also be used flexibly in the future.

> When should we pass the config file (like a kubeconfig)?

Please use `KUBECONFIG` environment variable. We don't really need to care about this cuz we can use `clientcmd.NewDefaultClientConfigLoadingRules()`.

See the client-go's documentation.
https://pkg.go.dev/k8s.io/client-go/tools/clientcmd

**Comment by 196Ikuchil** [comment]
Hi @sanposhiho @lianghao208 
I have finished the prework of this issue and it was merged.
So, I'd like to start the work on the remaining tasks.
If you have any messages, feel free to mention me on here👌

**Comment by sanposhiho** [comment]
/reopen

The remain task is the documentation.

**Comment by 196Ikuchil** [comment]
> The remain task is the documentation.

I'm waiting for the #106 issue to be resolved.

**Comment by pravarag** [comment]
Hi @196Ikuchil does the above PR also fixes #106 or it partially implements that? I'm also in way to add documentation as part of that issue.

**Comment by 196Ikuchil** [comment]
Hi, @pravarag Sorry for my confusing comment.
That PR doesn't fix issue #106 I think. It describes how to use the function of importing existing clusters. 
But, indeed, it also describes some environment variables.
So, could you add those environment variables to your documentation?
Thank you!

**Comment by pravarag** [comment]
> Hi, @pravarag Sorry for my confusing comment. That PR doesn't fix issue #106 I think. It describes how to use the function of importing existing clusters. But, indeed, it also describes some environment variables. So, could you add those environment variables to your documentation? Thank you!

Thanks for mentioning, I'll add the above ones as well.
